04-graphs/05-application.md

```diff
@@ -0,0 +1,704 @@
+
+`convert_maze_to_graph` iterates over each cell in the maze, and for each cell, it iterates over each of the four directions. The time complexity will be O(N) for the number of cells in the maze, and O(1) for the number of directions. The total time complexity will be O(N) * O(1), which simplifies to O(N).
+
+If the number of directions were somehow variable, this would look more like regular graph searching with O(N + E) time complexity, with E being the total number of cell connections. But since the number of directions is fixed, we can treat it as a constant.
+
+Since the input to `convert_maze_to_graph` is a 2-dimensional array, it can be tempting to think that the time complexity is O(N^2) since we use a nested loop to process the data, but this is not the case. The input is a 2-dimensional array, but the number of cells in the maze is not N^2. Rather, the number of cells is the number of rows times the number of columns, which is R * C, where R is the number of rows and C is the number of columns. So we could describe the time complexity as O(R * C), which is still not the same as O(N^2). But it's also fine to stick with calling N the total number of cells in the maze, which is R * C, and say that the time complexity is O(N). It really comes down to whether it's more useful to think about the problem in terms of the total amount of data (N), or the shape of the data (R * C).
+
+For the space complexity of `convert_maze_to_graph`, the 2-dimensional array is passed in, and therefore does not contribute to the space complexity. Our function works by building up a dictionary representing the graph, which will contain at most N nodes. Therefore, the space complexity will be O(N). Again, as with the time complexity, we could allow for the possibility of a variable number of directions, which would make the space complexity O(N + E), but since the number of directions is fixed, we can treat it as a constant.
+
+In total, for converting the maze to a graph and then searching the graph for a path, the time complexity will be O(N) + O(N), which simplifies to O(N). The space complexity will also be O(N) + O(N), which simplifies to O(N). Not too shabby!
+
+## Summary
+
+In this lesson, we applied the concepts of graphs to a real-world problem: finding a path through a maze. We started by representing the maze as a grid, and then converted it into a graph. We then used a recursive depth first search to find a path through the maze. Though the core of the algorithm remained the same, we needed to make some modifications to track the path as we searched the graph. When using graph searching to solve problems, it is very common to pass along some extra data to keep track of our progress towards a solution. The specific additional data and how we update it will vary according to what the problem is looking for.
```

Comment:
The rest of this looks great. The only thing I can think of that might be beneficial would to potentially have a small section for extension questions. Things like breadth first search, iterative depth first search and 3D mazes are mentioned a couple of times. It might be nice for the students to have one space where all the maze extensions are listed out in one place so they can see them all in one place. Kind of a "things to think about" section if you will. I can also see how that could be a little overwhelming for the students though, so It's fine if left out as well!

Reply:
I really wouldn't expect a student to come back to this to try the other approaches. I mostly had the earlier callout to give some kind of justification about why we went with the approach we did, rather than really having students dig further into it. So I think I'll leave this one as is.

Nothing to see here. Please move along. 😁